File: Example/javaapp/src/main/java/com/stripe/example/javaapp/viewmodel/LocationSelectionViewModel.java
package com.stripe.example.javaapp.viewmodel;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import com.stripe.example.javaapp.fragment.location.LocationListState;
import com.stripe.stripeterminal.Terminal;
import com.stripe.stripeterminal.external.callable.LocationListCallback;
import com.stripe.stripeterminal.external.models.ListLocationsParameters;
import com.stripe.stripeterminal.external.models.Location;
import com.stripe.stripeterminal.external.models.TerminalException;

import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.List;

final public class LocationSelectionViewModel extends ViewModel {
    private static final int LOCATION_QUERY_LIMIT = 100;

    final private MutableLiveData<LocationListState> mutableListState = new MutableLiveData<>(new LocationListState());
    public LiveData<LocationListState> listState = mutableListState;

    final private MutableLiveData<TerminalException> mutableError = new MutableLiveData<>();
    public LiveData<TerminalException> error = mutableError;

    final private LocationListCallback locationCallback = new LocationListCallback() {
        @Override
        public void onSuccess(@NotNull List<Location> locations, boolean hasMore) {
            LocationListState current = mutableListState.getValue();
            List<Location> combinedLocations = new ArrayList<>();
            combinedLocations.addAll(current.getLocations());
            combinedLocations.addAll(locations);
            LocationListState newState = new LocationListState(
                    combinedLocations,
                    false,
                    hasMore
            );
            mutableListState.postValue(newState);
        }

        @Override
        public void onFailure(@NotNull TerminalException e) {
            mutableError.postValue(e);
        }
    };

    public LocationSelectionViewModel() {
        super();
        load();
    }

    public void loadMoreLocations() {
        if (listState.getValue().isLoading()) {
            return;
        }
        if (!listState.getValue().hasMore()) {
            return;
        }
        LocationListState current = mutableListState.getValue();
        LocationListState newState = new LocationListState(
            current.getLocations(),
            true,
            current.hasMore()
        );

        mutableListState.postValue(newState);

        ListLocationsParameters.Builder builder = new ListLocationsParameters.Builder();
        builder.setLimit(LOCATION_QUERY_LIMIT);
        builder.setStartingAfter(current.getLocations().get(current.getLocations().size() - 1).getId());
        Terminal.getInstance().listLocations(builder.build(), locationCallback);
    }

    public void reload() {
        mutableListState.postValue(new LocationListState());
        load();
    }

    private void load() {
        ListLocationsParameters.Builder builder = new ListLocationsParameters.Builder();
        builder.setLimit(LOCATION_QUERY_LIMIT);
        Terminal.getInstance().listLocations(builder.build(), locationCallback);
    }
}

File: Example/kotlinapp/src/main/java/com/stripe/example/viewmodel/LocationSelectionViewModel.kt
package com.stripe.example.viewmodel

import androidx.lifecycle.ViewModel
import com.stripe.example.fragment.location.LocationListState
import com.stripe.stripeterminal.Terminal
import com.stripe.stripeterminal.external.callable.LocationListCallback
import com.stripe.stripeterminal.external.models.ListLocationsParameters
import com.stripe.stripeterminal.external.models.Location
import com.stripe.stripeterminal.external.models.TerminalException
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow

private const val LOCATION_QUERY_LIMIT = 100

class LocationSelectionViewModel : ViewModel() {
    private val mutableListState = MutableStateFlow(LocationListState())
    val listState: StateFlow<LocationListState> = mutableListState

    private val mutableError = MutableSharedFlow<TerminalException>()
    val error: Flow<TerminalException> = mutableError

    private val locationCallback = object : LocationListCallback {
        override fun onSuccess(locations: List<Location>, hasMore: Boolean) {
            mutableListState.value = mutableListState.value.let {
                it.copy(
                    locations = it.locations + locations,
                    hasMore = hasMore,
                    isLoading = false,
                )
            }
        }

        override fun onFailure(e: TerminalException) {
            mutableError.tryEmit(e)
        }
    }

    init {
        load()
    }

    fun loadMoreLocations() {
        if (listState.value.isLoading) return
        if (!listState.value.hasMore) return

        mutableListState.value = mutableListState.value.copy(isLoading = true)

        val parameters = ListLocationsParameters.Builder()
            .apply {
                startingAfter = mutableListState.value.locations.last().id
                limit = LOCATION_QUERY_LIMIT
            }
            .build()
        Terminal.getInstance().listLocations(parameters, locationCallback)
    }

    fun reload() {
        mutableListState.value = LocationListState()
        load()
    }

    private fun load() {
        Terminal.getInstance().listLocations(
            ListLocationsParameters.Builder().apply {
                limit = LOCATION_QUERY_LIMIT
            }.build(),
            locationCallback
        )
    }
}
